### DISCUSSION TITLE
[Q&A] Can 1 FL Participant act as both Server and Client in Vertical FL setting?

### DISCUSSION THREAD

**Discussion by apandit-algo** [NVIDIA/NVFlare#2376] [Q&A]
### Python version (`python3 -V`)

3.9.18

### NVFlare version (`python3 -m pip list | grep "nvflare"`)

2.4.0

### NVFlare branch (if running examples, please use the branch that corresponds to the NVFlare version, `git branch`)

2.4.0

### Operating system

Ubuntu 22.04

### Have you successfully run any of the following examples?

- [X] hello-numpy-sag with simulator
- [X] hello-pt with simulator
- [X] hello-numpy-sag with POC
- [X] hello-pt with POC

### Please describe your question

I am currently trying to follow the Vertical FL (SplitNN) [example](https://github.com/NVIDIA/NVFlare/tree/3460970cdff8ec9710742a8bcf3c0ccd5c1d267a/examples/advanced/vertical_federated_learning).
In this example, it uses the CIFAR-10 data but I wanted to try out VFL in NVFlare using a tabular data instead of CIFAR-10.

**As per the example, the federation setting is as below,**
|Participant|Labels|Features|
|-----|-----|-----|
|Server|Yes; Multiple Classes|No|
|Client 1|No|Yes; Images|

**The VFL scenario I am trying to achieve is as below,**
|Participant|Labels|Features|
|-----|-----|-----|
|Server|Yes; Binary Classes|10|
|Client 1|No|20|

**Query:** What I am trying to understand is, if it is possible in NVFlare for 1 participant to act as both the Server and Client? i.e. the participant will hold both the bottom model and the top model.

Refer to this scenario.
![image](https://github.com/NVIDIA/NVFlare/assets/143999400/90b24b97-365b-42f7-828b-9c4ab35d8168)

**Comment by YuanTingHsieh** [comment]
Hi @apandit-algo ,

Let's assume you have two participants, each has their own machine (P1 and P2).
You can just run the FL server and one FL client (client1) on machine P1.
And run the other FL client (client2) on machine P2.

Is this an OK setting for your use case?

**Reply by apandit-algo** [reply]
Yes, this would work for me. Currently I am trying to implement this same scenario.

Would like to ask a follow up question.
* Do I only need to change the job's `meta.json` file to point both the FL Server and FL Client to 1 machine?
**OR**
* Do I also need to change the underlying code in any Controllers, Learners, etc.?

**PS.:** I haven't yet started understanding the `production` mode, but testing everything using the `poc` mode.

**Reply by YuanTingHsieh** [reply]
Got it, add to @chesterxgchen 's answer.

For both of your questions, you DO NOT need to do anything to meta.json or the underlying code.

For the "production" mode, what you will need make sure that machine 2 can see machine 1 by its FQDN.
So if you machine 1 is running on AWS with a name "ubuntu@aws.west.com" then you need to modify the project.yml as follows:
(Make sure to open those ports on the machine 1 so that machine 2 can connect!)

```
participants:
  - name: ubuntu@aws.west.com
    type: server
    org: nvidia
    fed_learn_port: 8002
    admin_port: 8003
  - name: site-1
    type: client
    org: nvidia
  - name: site-2
    type: client
    org: nvidia
  - name: admin@nvidia.com
    type: admin
    org: nvidia
    role: project_admin
```

After "nvflare provision" command, you should have 1 server folder and 2 clients folder and 1 admin folder generated.
You need to put the server folder and site-1 folder into machine 1, start both of them.
You need to put the site-2 folder into machine 2, start it.
Then you can use the admin console to interact with your system.

Please refer to our documentation for more details:

https://nvflare.readthedocs.io/en/main/real_world_fl.html

**Reply by YuanTingHsieh** [reply]
@apandit-algo ^ does this help?

**Comment by chesterxgchen** [comment]
This should be ok for server and client running under same host ( kind like our current POC mode). server process and client1 process are just both use the same hostname

I don't think you need specify configuration for this case. FL server and Client 1 process  just run from different terminal

**Comment by YuanTingHsieh** [comment]
close due to inactivity, feel free to re-open
